Docs: revised join command (#353)
Redocly/openapi-cli: docs/commands/join.md

---
tocMaxDepth: 3
---
# `join`

## Introduction

:::warning Important
The `join` command is considered an experimental feature. This means it's still a work in progress and may go through major changes.

The `join` command supports OpenAPI 3.x definitions only.
:::

> With the `join` command, Redocly OpenAPI CLI can combine two or more API definition files into one so that maintainers of multiple API definitions can benefit from storing each endpoint as a standalone API definition file. However, this approach is not supported by the majority of OpenAPI tools, as they require a single API definition file.

Reviewer: These two sentences don't connect logically. In the original text, the meaning was something like this:

1) Maintainers sometimes split their endpoints into standalone files instead of keeping them in the root file.
2) Most OpenAPI tools don't work with standalone files like that.
3) The join command helps with this problem because it can combine those standalone files into one. 

In this new version, these things are out of order. The second sentence talks about the approach not being supported because other tools need a single file...but you already explained how to get the single file in the previous sentence? So this second sentence is just confusing.

You may want to rewrite this part so that you start with the problem first, and then explain how the join command can resolve it. Or you can start by describing what the join command does, and then explain in which use-cases it can be the most helpful.

To easily distinguish the origin of OpenAPI objects and properties, you can optionally instruct the `join` command to append custom prefixes to them.

The `join` command accepts both YAML and JSON files, which you can mix in the resulting `openapi.yaml` file. Setting a custom name for this file is not supported, which also means that if there is a file called `openapi.yaml` in the working directory, it will be overwritten.

Apart from providing individual API definition files as the input, you can also specify the path to a folder that contains multiple API definition files and match them with a wildcard (for example, `myproject/openapi/*.yaml`). The `join` command will collect all matching files and combine them into one file.

### Usage

```bash
openapi join <entrypoint> <entrypoint>...
openapi join <path-to-folder>/<wildcard-pattern> [--lint]
openapi join [--help] [--prefix-components-with-info-prop] [--prefix-tags-with-info-prop] [--prefix-tags-with-filename]

openapi join first-api.yaml second-api.yaml
openapi join first-api.yaml second-api.json
openapi join ./*.yaml
openapi join --version
```

## Options

> option                               | type               | required?    | default     | description

Reviewer: ```suggestion
Option                               | Type               | Required    | Default     | Description
```

-------------------------------------|:------------------:|:------------:|:-----------:|------------

> `entrypoints`                        | `array`, wildcard  | yes          | `[]`, -     | 1. Array of paths to API definition files that need to be joined. At least two input files are required.<br />2. A wildcard pattern to match API definition files within a specific folder

Reviewer: ```suggestion
`entrypoints`                        | `array`, wildcard  | yes          | `[]`, -     | 1. Array of paths to API definition files that you want to join. At least two input files are required.<br />2. A wildcard pattern to match API definition files within a specific folder
```

`--help`                             | `boolean`          | no           | -           | Show help
`--lint`                             | `boolean`          | no           | `false`     | Lint definition files
`--prefix-components-with-info-prop` | `string`           | no           | -           | Prefix components with property value from info object. See the [section](#) below
`--prefix-tags-with-info-prop`       | `string`           | no           | -           | Prefix tags with property value from info object. See the [section](#) below

> `--prefix-tags-with-filename`        | `boolean`          | no           | `false`     | Prefix tags with property value from file name. See the [section](#) below

Reviewer: These links don't lead anywhere - what should they link to? Also, please try using a more descriptive anchor text - if all sentences just say "See the section below", it could be difficult for the readers to distinguish which section you're talking about.

The `join` command accepts both YAML and JSON files, which you can mix in the resulting `openapi.yaml` file. Setting a custom name for this file is not supported, which also means that if there is a file called `openapi.yaml` in the working directory, it will be overwritten.

Apart from providing individual API definition files as the input, you can also specify the path to a folder that contains multiple API definition files and match them with a wildcard (for example, `myproject/openapi/*.yaml`). The `join` command will collect all matching files and combine them into one file.

### Usage

```bash
openapi join <entrypoint> <entrypoint>...
openapi join <path-to-folder>/<wildcard-pattern> [--lint]
openapi join [--help] [--prefix-components-with-info-prop] [--prefix-tags-with-info-prop] [--prefix-tags-with-filename]

openapi join first-api.yaml second-api.yaml
openapi join first-api.yaml second-api.json
openapi join ./*.yaml
openapi join --version
```

## Options

Option                               | Type               | Required     | Default     | Description
-------------------------------------|:------------------:|:------------:|:-----------:|------------

> `entrypoints`                        | `array`, wildcard  | yes          | `[]`, -     | 1. Array of paths to API definition files that you want to join. At least two input files are required.<br />2. A wildcard pattern to match API definition files within a specific folder

Reviewer: Question about the Default column - does this `[], - ` mean that there's no default value?

Author: Yes. It's better to leave a hyphen here

Reviewer: Got it, thanks!

    description: "Text from info: description of the second input file"
```

:::info
* If some operations in an input file don't have a tag assigned to them, the `join` command automatically adds the `other` tag to those operations in the output file. The `other` tag is also included in the `x-tagGroups` object.
* If any of the input files contain the `x-tagGroups` object, the content of this object is ignored by the `join` command and not included in the output file.
:::

The `servers` object combines the content from all input files, starting with the content from the first file. Commented lines are not included in the output file.

Path names and component names must be unique in all input files, but their content doesn't have to be unique for the `join` command to produce the output file. For each path item object, only the operation ID must be unique.

If the `join` command detects any conflicting content while trying to combine the input files, it displays informative messages about the conflicts and exits without creating an output file. To prevent this, use optional parameters to add prefixes to tags and components.

> ```bash example of conflicts

Reviewer: I really like that you included this example! Good thinking - it will help users recognize this type of error output, so they know what to expect. :brain:
